buildtest/buildsystem/base.py

```diff
@@ -266,6 +267,36 @@ def __init__(self, name, recipe, buildspec=None, testdir=None):
                 % (self.type, self.recipe.get("type"))
             )
 
+        self.shell = self.recipe.get("shell")
+        # if 'shell' key not found in Buildspec, let's set it to BUILDTEST_SHELL and log a message
+        if not self.shell:
+            self.logger.debug(
+                f"shell not found in Buildspec, setting shell to {BUILDTEST_SHELL} "
+            )
+            self.shell = BUILDTEST_SHELL
+
+        self.shell_name = self.shell.split()[0]
+        self.shell_opts = self.shell.split()[1:]
+
+        self.shebang = self.recipe.get("shebang")
+        # if shebang is not found in recipe let's set it to #!/bin/bash
+        if not self.shebang:
+            # when shell: python we set shebang to location of 'python' wrapper
+            if "python" in self.shell_name:
+                self.shebang = "#!" + shutil.which("python")
```

Comment:
This will trigger an error if Python isn’t found and returned as None.

Reply:
Wouldn’t there be a more elegant way to do this instead of hard coding python?

Reply:
Yea if python is not found we can raise an error. Also ``shell: python`` adapts to both python2 or python3 but it's not clear to user when they write in ``run`` section and test is not portable since it depends on user python environment. So if it works for me it may not necessarily work for you.

Right now we don't have ability to pass system python let's say ``shell: /usr/bin/python`` or ``shell: /usr/bin/python3`` similarly.

I particularly don't like having ``env`` support when shell is python, it causes some confusion to user when it sees buildtest modify the python script. I was thinking of removing this support but didnt for time being

Reply:
I don't think it's possible for ``python`` to not be found when running buildtest. It would cause an error running buildtest. In the ``master`` branch we had the **buildtest** https://github.com/buildtesters/buildtest/blob/master/bin/buildtest as a shell script that checks for the python client as a requirement. That has been removed in ``devel`` but i can see it being useful having these checks at start of buildtest.

Reply:
We only have python3 on sherlock, which is what made me think of this. I think (given that there can be multiple shells allowed) instead of hard coding them in specifically, it should be done with a single list, the schema validates, but then the code extracts (whatever the valid thing is). That's a much more elegant implementation than hard coding python into the code (and all the other future cases).

Reply:
Are you referring to modifying the schema line https://github.com/buildtesters/schemas/blob/master/script/script-v0.0.1.schema.json#L44 

Do you want add python3 and system python (/bin/python, /bin/python3) into the pattern

```
"pattern": "^(/bin/bash|/bin/sh|sh|bash|python|python3|/bin/python|/bin/python3).*"
```

The python interpreter will be whatever python wrapper was used when user is invoking buildtest so it's not possible to not have python in $PATH. I think having the distinction between python3 and python2 would be useful so user knows how to write their script in ``run`` section.

Reply:
I think I've mentioned this before - in that shell is an executable, I don't think we should limit the user to some known set. I should be able to run bash, sh (or a specific location, /bin/sh) or python3, or python4 when it comes out. We shouldn't need to pre-define the executables that the use can specify, only look for them with shutil.which() and exit with a clear error if they chose something not findable.

Reply:
If i understand you like to run ``shutil.which()`` on ``self.shell_name`` and if its ``None`` raise an error.  This would be much earlier in the code, but note that we have ``BuildTestCommand`` raise this error at https://github.com/buildtesters/buildtest/blob/devel/buildtest/utils/command.py#L46 which would eventually catch this exception.

Reply:
Yep.

Reply:
See https://github.com/buildtesters/buildtest/pull/290/commits/922756693cf0320010121c16e3deb243b2c756ba